tests/README.md

```diff
@@ -0,0 +1,1510 @@
+ilObject2 instance is created otherwise the expectations have no effect.
+
+```php
+<?php
+require_once './libs/composer/vendor/autoload.php';
+
+use Mockery\Adapter\Phpunit\MockeryPHPUnitIntegration;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * Class ilObject2Test
+ *
+ * @author  Nicolas Schäfli <ns@studer-raimann.ch>
+ *
+ * @runTestsInSeparateProcesses
+ * @preserveGlobalState    disabled
+ * @backupGlobals          disabled
+ * @backupStaticAttributes disabled
+ */
+class ilObject2Test extends TestCase {
+
+	use MockeryPHPUnitIntegration;
+	/**
+	 * @var ilObject2 $subject
+	 */
+	private $subject;
+	/**
+	 * @var ilObject | \Mockery\MockInterface $subjectParent
+	 */
+	private $subjectParent;
+
+	/**
+	 * @test
+	 * @small
+	 */
+	public function testCreateWithoutCloneModeWhichShouldSucceed() {
+
+		//expectations must be set before the child class is loaded
+		$expectedId = 5;
+		$this->subjectParent = Mockery::mock('overload:' . ilObject::class);
+		$this->subjectParent->shouldReceive('create')
+			->once()
+			->andReturn($expectedId);
+		$this->subject = Mockery::mock(ilObject2::class . '[]', []);
+
+		$result = $this->subject->create();
+
+		$this->assertSame($expectedId, $result);
+	}
+
+	//test other cases ...
+}
+```   
+In cases as shown above, each test has to be run in a separate PHP process because of the specially loaded classes. This can be done with the *@runTestsInSeparateProcesses* annotation.
+
+<a name="partial-mocks"></a>
+##### Partial mocks
+On the line with the suspicious looking string concatenation of the class name with
+empty brackets:
+
+```
+$this->subject = Mockery::mock(ilObject2::class . '[]', []);
+```
+
+This is directly creating a partial mock without any mocked methods. That is
+useful to test abstract classes without an actual implementation. The second parameter
+is used to pass additional constructor arguments.
+
+<a name="class-overload"></a>
+##### Class overload
+The string concatenation with overload string and the classname tells mockery to replace the class. This is possible with an autoloader trick which loads the mock class instead of the real implementation.
+
+```
+$this->subjectParent = Mockery::mock('overload:' . ilObject::class);
+```
+
+<a name="mock-static-calls"></a>
+#### Mock static calls
+In some situation it is necessary to mock an entire class due to static method access or a new call.
+Mockery provides a convenient way to load a class alias which replaces the class entirely. The following
+example illustrates mocking of static method calls to test the Stream class within the filesystem service.
+
+The concrete problem within the filesystem service was that this service has to interact with php built in
+functions to manipulate the underlying resource of the stream. PHP has no functionality to autoload functions,
+which makes them difficult to mock. The solution of the author was to wrap the PHP functions with a
+helper class which can be replaced.
```

Comment:
A better solution would be to avoid static calls at all. This is generally suprisingly simple (see examples above) and avoids these problems when unit testing. Names for concrete implementations are just bad, no matter if they belong to classes, static methods or functions. From the perspective of unit testing and composability a call to a named function is in no way better than a call to a static method.

Reply:
I agree that calls to static method or functions should be avoided in most cases. However in terms of testability are static method calls still a bit better because you can replace them while autoloading (which is as already mentioned above not a good solution) but function are not autoloaded at all which make them even harder. The only way to "mock" a function is to create an identical one which is closer to the class which is using the function (namespace).

But the optimal solution would be to make us of the DIP of course.

Reply:
As you are proposing a better solution here ("use static calls instead of functions") you might as well propose the even better solution ("use methods of service objects instead of static calls or functions").